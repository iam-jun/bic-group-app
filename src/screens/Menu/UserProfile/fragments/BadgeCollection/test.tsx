import React from 'react';
import {
  View, Dimensions, FlatList, StyleSheet,
} from 'react-native';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import Avatar from '~/baseComponents/Avatar';
import Icon from '~/baseComponents/Icon';
import spacing, { borderRadius } from '~/theme/spacing';

const Grid = ({ data }: {data: any[]}) => {
  const deviceWidth = Dimensions.get('window').width;

  const itemWidth = 48; // Constant item width
  const numColumns = Math.floor((deviceWidth - 10 * 2) / (itemWidth + 6 * 2));
  const rowWidth = numColumns * (itemWidth + 6 * 2);
  const paddingLeft = (deviceWidth - (10 * 2) - rowWidth) / 2;

  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;
  const styles = themeStyles(theme);

  const isSelected = false;

  const renderGridItem = ({ item, index }: any) =>
    // Render each grid item here
    (
      <View
        key={index}
        style={{
          margin: 6,
          borderWidth: isSelected ? 2 : 1,
          borderColor: isSelected ? colors.purple50 : colors.neutral5,
          borderRadius: borderRadius.pill,
        }}
      >
        <Avatar.Medium
          isRounded
          source={{ uri: item?.iconUrl }}
        />
        {isSelected
          ? (
            <View style={styles.iconChose}>
              <Icon size={10} icon="Check" tintColor={colors.white} />
            </View>
          )
          : null}
      </View>
    );

  return (
    <View style={{
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      paddingHorizontal: 10,
      paddingBottom: 10,
    }}
    >

      <View style={{
        // paddingLeft,
      }}
      >
        <FlatList
          data={data}
          renderItem={renderGridItem}
          keyExtractor={(item, index) => index.toString()}
          numColumns={numColumns}
        />
      </View>
    </View>
  );
};

const themeStyles = (theme: ExtendedTheme) => {
  const { colors } = theme;

  return StyleSheet.create({
    container: {
      flex: 1,
      paddingTop: spacing.padding.large,
    },
    iconChose: {
      backgroundColor: colors.purple50,
      borderRadius: borderRadius.pill,
      height: 20,
      width: 20,
      alignItems: 'center',
      justifyContent: 'center',
      position: 'absolute',
      zIndex: 1,
      right: 0,
    },
  });
};

export default Grid;
