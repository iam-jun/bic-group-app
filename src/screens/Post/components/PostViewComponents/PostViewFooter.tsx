import React, { FC } from 'react';
import { StyleSheet, View } from 'react-native';
import { useDispatch } from 'react-redux';
import { ExtendedTheme, useTheme } from '@react-navigation/native';

import Button from '~/beinComponents/Button';
import EmojiBoard from '~/beinComponents/emoji/EmojiBoard';
import { IReactionCounts } from '~/interfaces/IPost';
import { IconType } from '~/resources/icons';
import * as modalActions from '~/store/modal/actions';
import { validateReactionCount } from './helper';
import dimension from '~/theme/dimension';
import Text from '~/beinComponents/Text';
import { useBaseHook } from '~/hooks';
import { spacing } from '~/theme';

export interface PostViewFooterProps {
  labelButtonComment: string;
  onAddReaction?: any;
  btnReactTestID?: string;
  btnCommentTestID?: string;
  onPressComment?: () => void;
  reactionCounts: IReactionCounts;
  canComment?:boolean;
  canReact?: boolean;
}

const PostViewFooter: FC<PostViewFooterProps> = ({
  labelButtonComment,
  onAddReaction,
  btnReactTestID,
  btnCommentTestID,
  onPressComment,
  reactionCounts,
  canComment,
  canReact,
}: PostViewFooterProps) => {
  const dispatch = useDispatch();
  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;
  const styles = createStyle(theme);
  const { t } = useBaseHook();

  const validReactionCount = validateReactionCount(reactionCounts);
  const numberOfReactions = Object.keys(reactionCounts).length;
  const labelReactionCount = `${
    numberOfReactions ? `${numberOfReactions} ` : ''
  }${t('post:button_react')}`;

  const onEmojiSelected = (
    emoji: string, key?: string,
  ) => {
    dispatch(modalActions.hideModal());
    if (key) {
      onAddReaction?.(key);
    }
  };

  const onPressReact = () => {
    const payload = {
      isOpen: true,
      ContentComponent: (
        <EmojiBoard
          width={dimension.deviceWidth}
          height={280}
          onEmojiSelected={onEmojiSelected}
        />
      ),
    };
    dispatch(modalActions.showModal(payload));
  };

  const renderReactButtonItem = (
    title: string,
    icon: IconType,
    onPress: any,
    onLongPress: any,
    disabled?: boolean,
    testID?: string,
  ) => (
    <View style={styles.buttonReactContainer}>
      <Button
        testID={testID}
        useI18n
        onPress={onPress}
        onLongPress={onLongPress}
        disabled={disabled}
        leftIcon={icon}
        leftIconProps={{
          icon,
          size: 20,
          tintColor: colors.neutral40,
        }}
        textProps={{
          variant: 'bodyM',
          color: colors.neutral80,
        }}
        style={styles.buttonReact}
      >
        {title}
      </Button>
    </View>
  );

  return (
    <View style={[styles.reactButtons, !canComment && !canReact && styles.disbaledReactComment]}>
      {(validReactionCount && !!canReact) && (
        <>
          {renderReactButtonItem(
            labelReactionCount,
            'iconReact',
            onPressReact,
            onPressReact,
            false,
            btnReactTestID,
          )}
        </>
      )}
      {!!canComment
       && renderReactButtonItem(
         labelButtonComment,
         'MessageDots',
         onPressComment,
         onPressComment,
         !onPressComment,
         btnCommentTestID,
       )}
      {!canComment && !canReact && (
      <View style={[styles.emptyView, styles.disbaledReactComment]}>
        <Text.BodyS color={theme.colors.neutral20} useI18n>post:text_cannot_comment_and_react</Text.BodyS>
      </View>
      )}
    </View>
  );
};

const createStyle = (theme: ExtendedTheme) => {
  const { colors } = theme;
  return StyleSheet.create({
    reactButtons: {
      flexDirection: 'row',
      height: dimension?.commentBarHeight,
      alignItems: 'center',
    },
    disbaledReactComment: {
      height: 36,
      marginTop: spacing.margin.tiny,
    },
    buttonReactContainer: {
      flex: 1,
      height: 'auto',
    },
    buttonReact: {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
    },
    emptyView: {
      backgroundColor: colors.neutral2,
      flex: 1,
      alignItems: 'center',
      justifyContent: 'center',
    },
  });
};

export default PostViewFooter;
