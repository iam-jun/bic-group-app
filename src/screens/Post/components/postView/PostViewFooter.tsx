import React, {FC} from 'react';
import {View, StyleSheet, Platform} from 'react-native';
import {useTheme} from 'react-native-paper';

import {ITheme} from '~/theme/interfaces';

import Divider from '~/beinComponents/Divider';
import {IconType} from '~/resources/icons';
import Div from '~/beinComponents/Div';
import Button from '~/beinComponents/Button';
import * as modalActions from '~/store/modal/actions';
import {useDispatch} from 'react-redux';
import EmojiBoard from '~/beinComponents/emoji/EmojiBoard';

export interface PostViewFooterProps {
  labelButtonComment: string;
  onAddReaction?: any;
  btnReactTestID?: string;
  btnCommentTestID?: string;
  onPressComment?: () => void;
}

const PostViewFooter: FC<PostViewFooterProps> = ({
  labelButtonComment,
  onAddReaction,
  btnReactTestID,
  btnCommentTestID,
  onPressComment,
}: PostViewFooterProps) => {
  const dispatch = useDispatch();
  const theme = useTheme() as ITheme;
  const {colors, spacing, dimension} = theme;
  const styles = createStyle(theme);

  const onEmojiSelected = (emoji: string, key?: string) => {
    dispatch(modalActions.hideModal());
    if (key) {
      onAddReaction?.(key);
    }
  };

  const onPressReact = (event: any) => {
    const payload = {
      isOpen: true,
      ContentComponent: (
        <EmojiBoard
          width={Platform.OS === 'web' ? 320 : dimension.deviceWidth}
          height={280}
          onEmojiSelected={onEmojiSelected}
        />
      ),
      props: {
        webModalStyle: {minHeight: undefined},
        isContextMenu: true,
        position: {x: event?.pageX, y: event?.pageY},
        side: 'center',
      },
    };

    if (Platform.OS !== 'web') {
      dispatch(modalActions.showModal(payload));
      return;
    }

    // Handling show reaction bottom sheet on web
    // @ts-ignore
    event.target.measure((fx, fy, width, height, px, py) => {
      const buttonReactPaddingBottom = spacing.padding.tiny || 4;
      let x = px;
      let y = py + height + buttonReactPaddingBottom;

      /*
      As target may be the label, not the whole button itself,
      which causes the bottom sheet will render in the middle.
      If pressing on the label, the childElementCount will equal 0.
      */
      if (event.target.childElementCount !== 0) {
        x = x + width / 2;
      } else {
        // Move menu further down when pressing on label
        y = y + buttonReactPaddingBottom * 1.5;
      }
      payload.props.position = {x, y};

      dispatch(modalActions.showModal(payload));
    });
  };

  const renderReactButtonItem = (
    title: string,
    icon: IconType,
    onPress: any,
    onLongPress: any,
    disabled?: boolean,
    testID?: string,
  ) => {
    return (
      <Div
        className="button-react"
        style={Platform.OS !== 'web' ? styles.buttonReactContainer : {}}>
        <Button
          testID={testID}
          useI18n
          onPress={onPress}
          onLongPress={onLongPress}
          disabled={disabled}
          leftIcon={icon}
          leftIconProps={{
            icon: icon,
            size: 14,
            tintColor: colors.textSecondary,
          }}
          textProps={{
            variant: 'bodySM',
            color: colors.textSecondary,
          }}
          style={styles.buttonReact}>
          {title}
        </Button>
      </Div>
    );
  };

  return (
    <View style={styles.reactButtons}>
      {renderReactButtonItem(
        'post:button_react',
        'iconReact',
        onPressReact,
        onPressReact,
        false,
        btnReactTestID,
      )}
      <Divider style={{height: '66%', alignSelf: 'center'}} horizontal />
      {renderReactButtonItem(
        labelButtonComment,
        'CommentAltDots',
        onPressComment,
        onPressComment,
        !onPressComment,
        btnCommentTestID,
      )}
    </View>
  );
};

const createStyle = (theme: ITheme) => {
  const {colors, dimension} = theme;
  return StyleSheet.create({
    container: {},
    reactButtons: {
      flexDirection: 'row',
      height: dimension?.commentBarHeight,
      borderTopWidth: 1,
      borderColor: colors.borderDivider,
      alignItems: 'center',
    },
    buttonReactContainer: {
      flex: 1,
      height: 'auto',
    },
    buttonReact: {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
    },
  });
};

export default PostViewFooter;
