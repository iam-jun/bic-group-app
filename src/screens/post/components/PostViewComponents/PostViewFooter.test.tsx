import React from 'react';
import { cleanup, renderWithRedux, store } from '~/test/testUtils';
import PostViewFooter from './PostViewFooter';

afterEach(() => {
  cleanup();
  store.clearActions();
});

describe('PostViewFooter component', () => {
  it('should render correctly', () => {
    const rendered = renderWithRedux(
      <PostViewFooter labelButtonComment="Comment" reactionCounts={{}} />,
    );
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
