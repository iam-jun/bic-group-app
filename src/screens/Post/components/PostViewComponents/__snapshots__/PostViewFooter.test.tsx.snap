// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostViewFooter component should render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "borderColor": "#EAEDF2",
      "borderTopWidth": 1,
      "flexDirection": "row",
      "height": 44,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "height": "auto",
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
        testID="button_wrapper.content"
      >
        <View
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                Object {
                  "marginHorizontal": 8,
                },
                Object {
                  "backgroundColor": undefined,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              <View
                testID="button_wrapper.icon"
              >
                <View
                  onLayout={[Function]}
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "position": "relative",
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "opacity": 1,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "#f9f9f9",
                          "height": 14,
                          "justifyContent": "center",
                          "width": 14,
                        }
                      }
                      testID="Image__placeholder"
                    />
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Object {
                          "height": 14,
                          "width": 14,
                        },
                      ]
                    }
                  >
                    <FastImageView
                      onFastImageError={[Function]}
                      onFastImageLoadEnd={[Function]}
                      resizeMode="cover"
                      size={14}
                      source={null}
                      style={
                        Object {
                          "bottom": 0,
                          "left": 0,
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#2F3945",
                "fontFamily": "OpenSans-SemiBold",
                "fontSize": 14,
              },
              Object {
                "color": "#627286",
              },
              Object {
                "textAlign": "center",
              },
            ]
          }
        >
          React
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "#EAEDF2",
          "height": undefined,
          "marginVertical": 0,
          "width": 1,
        },
        Object {
          "alignSelf": "center",
          "height": "66%",
        },
      ]
    }
    testID="divider"
  />
  <View
    style={
      Object {
        "flex": 1,
        "height": "auto",
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
        testID="button_wrapper.content"
      >
        <View
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                Object {
                  "marginHorizontal": 8,
                },
                Object {
                  "backgroundColor": undefined,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              <View
                testID="button_wrapper.icon"
              >
                <RNSVGSvgView
                  align="xMidYMid"
                  bbHeight={14}
                  bbWidth={14}
                  color={4284641926}
                  focusable={false}
                  height={14}
                  meetOrSlice={0}
                  minX={0}
                  minY={0}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderWidth": 0,
                      },
                      Object {},
                      Object {
                        "flex": 0,
                        "height": 14,
                        "width": 14,
                      },
                    ]
                  }
                  tintColor={4284641926}
                  vbHeight={24}
                  vbWidth={24}
                  width={14}
                >
                  <RNSVGGroup>
                    <RNSVGPath
                      d="M12 9a1 1 0 101 1 1 1 0 00-1-1zm7-7H5a3 3 0 00-3 3v10a3 3 0 003 3h11.59l3.7 3.71A1 1 0 0021 22a.84.84 0 00.38-.08A1 1 0 0022 21V5a3 3 0 00-3-3zm1 16.59l-2.29-2.3A1 1 0 0017 16H5a1 1 0 01-1-1V5a1 1 0 011-1h14a1 1 0 011 1zM8 9a1 1 0 101 1 1 1 0 00-1-1zm8 0a1 1 0 101 1 1 1 0 00-1-1z"
                      fill={
                        Array [
                          2,
                        ]
                      }
                      propList={
                        Array [
                          "fill",
                        ]
                      }
                    />
                  </RNSVGGroup>
                </RNSVGSvgView>
              </View>
            </View>
          </View>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#2F3945",
                "fontFamily": "OpenSans-SemiBold",
                "fontSize": 14,
              },
              Object {
                "color": "#627286",
              },
              Object {
                "textAlign": "center",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
